backup/moodle2/backup_format_vsf_plugin.class.php

```diff
@@ -0,0 +1,134 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+/**
+ * Specialised backup for Progress Section course format.
+ *
+ * @package   format_vsf
+ * @category  backup
+ * @copyright &copy; 2022-onwards G J Barnard in respect to modifications of standard topics format.
+ * @copyright &copy; 2024-onwards RvD in respect to modifications for custom icons.
+ * @copyright 2017 Marina Glancy
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+/**
+ * Specialised backup for Progress Section course format.
+ *
+ * Processes 'numsections' from the old backup files and hides sections that used to be "orphaned".
+ *
+ * @package   format_vsf
+ * @category  backup
+ * @copyright &copy; 2022-onwards G J Barnard in respect to modifications of standard topics format.
+ * @copyright &copy; 2024-onwards RvD in respect to modifications for custom icons.
+ * @copyright 2017 Marina Glancy
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class backup_format_vsf_plugin extends backup_format_plugin {
+
+    public function __construct($plugintype, $pluginname, $optigroup, $step) {
+        parent::__construct($plugintype, $pluginname, $optigroup, $step);
+    }
+
+    /**
+     * Define course plugin structure for format_vsf
+     *
+     * @return array|void
+     */
+    protected function define_course_plugin_structure() {
+        global $DB;
+        // Define the virtual plugin element with the condition to fulfill.
+        $plugin = $this->get_plugin_element(null, $this->get_format_condition(), 'vsf');
+
+        // Create one standard named plugin element (the visible container).
+        // The courseid not required as populated on restore.
+        $pluginwrapper = new backup_nested_element($this->get_recommended_name());
+        // Connect the visible container ASAP.
+        $plugin->add_child($pluginwrapper);
+
+        // We'll simply do ALL installed modules.
+        // It will be too much hassle to see what we may wish to include or not,
+        // because this is "high level behaviour". Customizing icons also will not
+        // see what may or may not be wishful. Custom icons is NOT dependent on
+        // which modules are in use in the course.
+        // Customisation can be done for ALL activity types.
+        $courseiconwrapper = new backup_nested_element('courseicons');
+        $courseicon = new backup_nested_element('courseicon', [], ['name']);
+        $sql = 'SELECT DISTINCT m.name FROM {modules} m';
```

Comment:
Do all supported databases support 'DISTINCT'?  Would it not be better to use a DB API call instead of pure SQL as recommended (I believe I've seen this in the Moodle developer documentation)?